src/v/config/constraint_method_generators.cc

```diff
@@ -0,0 +1,561 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "config/constraint_method_generators.h"
+
+#include "cluster/types.h"
+#include "config/configuration.h"
+#include "model/fundamental.h"
+#include "vassert.h"
+
+#include <seastar/util/log.hh>
+
+#include <algorithm>
+
+namespace config {
+inline ss::logger logger{"constraints"};
+
+namespace {
+
+template<typename RangeMinT, typename T>
+bool less_than_min(const std::optional<RangeMinT> min, const T& topic_val) {
+    return min && topic_val < static_cast<T>(*min);
+}
+
+template<typename RangeMaxT, typename T>
+bool greater_than_max(const std::optional<RangeMaxT> max, const T& topic_val) {
+    return max && topic_val > static_cast<T>(*max);
+}
+
+template<typename RangeT, typename T>
+bool is_invalid_range(const range_values<RangeT>& range, const T& topic_val) {
+    return less_than_min<RangeT, T>(range.min, topic_val)
+           || greater_than_max<RangeT, T>(range.max, topic_val);
+}
+
+template<typename RangeT>
+range_values<RangeT>
+get_range(const constraint_args& args, const model::topic& topic) {
+    // NOTE to reviewers: Not using visit here or else need to check for
+    // signed/unsigned int comparison which is avoidable.
+    try {
+        return std::get<range_values<RangeT>>(args.flags);
+    } catch (const std::bad_variant_access& ex) {
+        // This should not happen since the variant is hardcoded in
+        // constraint_types.cc, but just-in-case, let's catch and log.
+        logger.error(
+          "Constraints failure[min/max not found]: topic property {}.{}",
+          topic(),
+          args.name);
+        return range_values<RangeT>{};
+    }
+}
```

Comment:
nit/suggestion: Maybe return a `std::optional`?

Reply:
Removed in the recent push.